charts/generic-prometheus-alerts/ci/tests/ingress-alerts-tests.yaml

```diff
@@ -0,0 +1,170 @@
+---
+rule_files:
+  - ../ingress-alerts.yaml
+
+evaluation_interval: 1m
+
+tests:
+  - interval: 1m
+    input_series:
+      - series: 'nginx_ingress_controller_requests{exported_namespace="test-application-dev", ingress="test-application-ingress", status="500"}'
+        values: '0 1 3'
+      - series: 'nginx_ingress_controller_requests{exported_namespace="test-application-dev", ingress="test-application-ingress", status="501"}'
+        values: '0 2 5'
+
+    name: 5xxErrorResponses alert fired as request rate increases over the duration of the 1 minute time range window
+
+    promql_expr_test:
+      # assert that
+      #   at 2m the rate (increase per second) in the last minute: for 500 requests it's 0.03 ((3 - 1) / 60), for 501 requests it's 0.05 ((5 - 2) / 60)
+      #   which is greater than 0
+      #   the average of those is 0.041
+      - expr: avg(rate(nginx_ingress_controller_requests{exported_namespace="test-application-dev", ingress="test-application-ingress", status=~"5.*"}[1m]) > 0) by (ingress, exported_namespace)
+        eval_time: 2m
+        exp_samples:
+          - labels: '{exported_namespace="test-application-dev", ingress="test-application-ingress"}'
+            value: 0.04166666666666667
+
+    alert_rule_test:
+      # assert that the 5xx alert is firing
+      - eval_time: 2m
+        alertname: 5xxErrorResponses
+        exp_alerts:
+          - exp_labels:
+              alertname: 5xxErrorResponses
+              exported_namespace: test-application-dev
+              ingress: test-application-ingress
+              severity: alert-severity-tag
+            exp_annotations:
+              message: "Ingress test-application-dev/test-application-ingress is serving 5xx responses."
+              runbook_url: https://runbook.com/ingress-5xx-error-responses
+              dashboard_url: https://grafana.live.cloud-platform.service.justice.gov.uk/d/golden-signals/golden-signals?orgId=1&var-namespace=test-application-dev&var-service=test-application
+
+  - interval: 1m
+    input_series:
+      - series: 'nginx_ingress_controller_requests{exported_namespace="test-application-dev", ingress="test-application-ingress", status="500"}'
+        values: '0 1 0'
+
+    name: 5xxErrorResponses alert not fired as request rate decreases over the duration of the 1 minute time range window
+
+    promql_expr_test:
```

Comment:
I found the inclusion of these tests quite confusing... having looked at your nice document I realised that you were kind-of using them to further your understanding, which is good.

But i would still like to see something that distinguishes these `promql-expr-test` tests from the "real" `alert_rule_test` tests - maybe a different file? Or maybe just a comment above each one might be enough?

Reply:
Good shout - addressed with https://github.com/ministryofjustice/hmpps-helm-charts/pull/79/commits/2600d5300d673ba8f3900e60d9886805c48c4fe3. Hopefully the comment helps?